references/cli/kube.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"k8s.io/kubectl/pkg/util/templates"
+	"k8s.io/utils/strings/slices"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/oam-dev/kubevela/apis/types"
+	velacmd "github.com/oam-dev/kubevela/pkg/cmd"
+	cmdutil "github.com/oam-dev/kubevela/pkg/cmd/util"
+	"github.com/oam-dev/kubevela/pkg/multicluster"
+	"github.com/oam-dev/kubevela/pkg/utils"
+	"github.com/oam-dev/kubevela/pkg/utils/util"
+)
+
+// KubeCommandGroup command group for native resource management
+func KubeCommandGroup(f velacmd.Factory, streams util.IOStreams) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "kube",
+		Short: i18n.T("Managing native Kubernetes resources across clusters."),
+		Annotations: map[string]string{
+			types.TagCommandType: types.TypeCD,
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+
+		},
+	}
+	cmd.AddCommand(NewKubeApplyCommand(f, streams))
+	return cmd
+}
+
+// KubeApplyOptions options for kube apply
+type KubeApplyOptions struct {
+	files     []string
+	clusters  []string
+	namespace string
+	dryRun    bool
+
+	filesData []utils.FileData
+	objects   []*unstructured.Unstructured
+
+	util.IOStreams
+}
+
+// Complete .
+func (opt *KubeApplyOptions) Complete(f velacmd.Factory, cmd *cobra.Command) error {
+	opt.namespace = velacmd.GetNamespace(f, cmd)
+	opt.clusters = velacmd.GetClusters(cmd)
+
+	var paths []string
+	for _, file := range opt.files {
+		path := strings.TrimSpace(file)
+		if !slices.Contains(paths, path) {
+			paths = append(paths, path)
+		}
+	}
+	for _, path := range paths {
+		data, err := utils.LoadDataFromPath(cmd.Context(), path, utils.IsJSONOrYAMLFile)
+		if err != nil {
+			return err
+		}
+		opt.filesData = append(opt.filesData, data...)
+	}
+	return nil
+}
+
+// Validate .
+func (opt *KubeApplyOptions) Validate() error {
+	if len(opt.filesData) == 0 {
+		return fmt.Errorf("not file found")
```

Comment:
Should the length of `opt.files` be checked in Complete?

Reply:
good idea

Reply:
added a file length check in Validate, not in Complete.